python/thunder/rdds/series.py

```diff
@@ -883,6 +885,229 @@ def toSpatialSeries(self):
         return SpatialSeries(self.rdd).__finalize__(self)
 
+    def _makeMasks(self, index=None, level=0):
+        """
+        Internal function for generating masks for selecting values based on multi-index values
+    
+        As all other multi-index functions will call this function, basic type-checking is also
+        performed at this stage.
+        """
+
+        if index is None:
+            index = self.index
+        
+        try:
+            dims = len(array(index).shape)
+            if dims == 1:
+                index = array(index, ndmin=2).T
+        except:
+            raise TypeError('for multi-index functionality: index must be convertible to a numpy ndarray')
+
+        try:
+            index = index[:, level]
+        except:
+            raise ValueError("levels must be indices into individual elements of the index")
+
+        lenIdx = index.shape[0]
+        nlevels = index.shape[1]
+
+        combs = product(*[unique(index.T[i,:]) for i in xrange(nlevels)])
+        combs = array([l for l in combs])
+
+        masks = array([[array_equal(index[i], c) for i in xrange(lenIdx)] for c in combs])
+
+        return zip(*[(masks[x], combs[x]) for x in xrange(len(masks)) if masks[x].any()])
+
+    def _applyByIndex(self, level=None, function=None):
+        """
+        An internal function for applying a function to groups of values base on a multi-index
+
+        Elements of each record are grouped according to unique value combinations of the multi-
+        index across the given levels of the multi-index. Then the given function is applied
+        to to each of these groups separately. If this function is many-to-one, the result
+        can be recast as a Series indexed by the unique index values used for grouping.
+        """
+
+        if type(level) is int:
+            level = [level]
+
+        masks, ind = self._makeMasks(index=self.index, level=level)
+        bcMasks = self.rdd.ctx.broadcast(masks)
+        bcNum = self.rdd.ctx.broadcast(len(masks))
+        def f(data):
+            return array([function(data[bcMasks.value[x]]) for x in xrange(bcNum.value)])
+        rdd = self.rdd.mapValues(f)
+        index = array(ind)
+        if len(index[0]) == 1:
+            index = ravel(index)
+        return Series(rdd, index=index).__finalize__(self, noPropagate=('_dtype', '_index'))
+
+    def selectByIndex(self, level=0, val=None, squeeze=False, filter=False):
+        """
+        Select or filter elements of the Series by index values (across levels, if multi-index)
+
+        At each of the given levels, only values with a multi-index that matches one of the given
+        values will be retained for the returned Series. If filtering, this is inverted and only 
+        those elements will be ddropped.
+
+        Parameters:
+        -----------
+        level: An integer or an list of integers specificying which levels in the multilevel index to use when
```

Comment:
For parameter formatting in docstrings, the first line should be `argname : type`, as in `level: integer or list`, as well as a note about whether it's optional and what the default is, like `val: boolean, optional, default = False`. Then the second line should be a description like what you have here, but without restating the type to avoid redundancy.